src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs

```diff
@@ -86,5 +89,33 @@ public static int ExpandPrime(int oldSize)
 
             return GetPrime(newSize);
         }
+
+        // Takes an index and distributes it in the range [0, range)
+        // e.g. 0 to range exclusive.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint FastRange(uint index, uint range)
+        {
+            // Sse42.IsSupported check is baked into the R2R image so while the value returned
+            // from this method will be different on both branches, it will remain stable during
+            // a process' lifetime even if it is recompiled by tiered compilation.
+            if (Sse42.IsSupported)
```

Comment:
@tannergooding this is correct?

Reply:
Yes, that should be correct. S.P.Corelib will go through R2R which will bake in `SSE42.IsSupported` check as a call (which lookups the cached CPUID value). So if the CPU supports it, the true path will still be hit.
* For x86, it (R2R) will currently treat SSE/SSE2 as "always supported", ISAs dependent on the VEX encoding (or ISAs for other platforms, like ARM) as "never supported", and other ISAs (like `SSE42`) as runtime checks.

If this method becomes labeled as "hot", it will be rejitted and the normal JIT rules apply; which means `Sse42.IsSupported` will become a runtime constant and the same path will be hit, but this time with the dead path optimized away.

@jkotas and/or @MichalStrehovsky might have some comments on if there are any scenarios where this can fail (e.g. are there any scenarios where R2R or another AOT scenario won't treat this as a runtime call and where rejit can still happen).

Reply:
Thinking a bit more, one scenario where it could possibly fail is if the user explicitly sets `COMPlus_EnableSSE42=0`. This is a "retail" flag (available in `release` builds), but which is meant for validating that your fallback code paths work, so it is not necessarily for "production" use).

I think, in this case, R2R would still apply by default but rejit could then treat `Sse42` as "not supported", so it might be worth validating and documenting....

Reply:
I assume/hope the embedded `Sse42:get_IsSupported():bool` would return false in that circumstance to match the environment variable?
```asm
       call     [Sse42:get_IsSupported():bool]
       test     al, al
       je       SHORT G_M59265_IG10
       mov      eax, 0xD1FFAB1E
       crc32    eax, r12d
       mov      edx, r13d
       imul     rax, rdx
       shr      rax, 32
       mov      ecx, eax
       jmp      SHORT G_M59265_IG11

G_M59265_IG10:		;; bbWeight=0.50
       mov      eax, r12d
       xor      rdx, rdx
       div      edx:eax, r13d
       mov      ecx, edx
```

Reply:
I do not see a correctness problem here.